sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala

```diff
@@ -0,0 +1,151 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalyst.optimizer
+
+import org.apache.spark.sql.catalyst.expressions._
+import org.apache.spark.sql.catalyst.plans.logical._
+import org.apache.spark.sql.internal.SQLConf
+import org.apache.spark.sql.types._
+
+/**
+ * This aims to handle a nested column aliasing pattern inside the `ColumnPruning` optimizer rule.
+ * If a project or its child references to nested fields, and not all the fields
+ * in a nested attribute are used, we can substitute them by alias attributes; then a project
+ * of the nested fields as aliases on the children of the child will be created.
+ */
+object NestedColumnAliasing {
+
+  def unapply(plan: LogicalPlan)
+    : Option[(Map[GetStructField, Alias], Map[ExprId, Seq[Alias]])] = plan match {
+    case Project(projectList, child)
+        if SQLConf.get.nestedSchemaPruningEnabled && canProjectPushThrough(child) =>
```

Comment:
This rule applies even Spark can't prune nested fields. For example, when reading from data sources don't support nested schema pruning.

This is a special rule that only makes sense when pruning nested fields. But now it is applied to all queries. @dongjoon-hyun WDYT? Shall we apply it when pruning nested fields happens actually?

Reply:
Thank you for reivew, @viirya !
1. Do you mean JSON and AVRO?
> For example, when reading from data sources don't support nested schema pruning.
2. For the following,
> Shall we apply it when pruning nested fields happens actually?

This line is `unapply` and this is used [at the following pattern](https://github.com/apache/spark/pull/23964/files#diff-a636a87d8843eeccca90140be91d4fafR650) inside `ColumnPruning` optimizer rule and `NestedColumnAliasing.replaceToAliases` is not called if `getAliasSubMap`
```scala
    case p @ NestedColumnAliasing(nestedFieldToAlias, attrToAliases) =>
	      NestedColumnAliasing.replaceToAliases(p, nestedFieldToAlias, attrToAliases)
```

Reply:
> 1. Do you mean JSON and AVRO?

There might be data sources not included in Spark and they don't support nested schema pruning.

> This line is unapply and this is used at the following pattern inside ColumnPruning optimizer rule and NestedColumnAliasing.replaceToAliases is not called if getAliasSubMap return None.

`getAliasSubMap` returns something when the config is enabled and nested field access is found. But nested schema pruning works when underlying data source supports it. Otherwise, pushing down and aliasing nested field access is not useful. It doesn't have negative effect in query execution though. Maybe not a big deal, it is probably better if we apply this rule when nested schema pruning actually happens. It should not block this, I think.

Reply:
Yep. I don't think that block this.

Reply:
Agree with @viirya , this rule can be useful even for datasources that do't support nested pruning. We can have a followup to make it as a separate conf to enable it.